script: Add support for parsing CSS in parallel

mrobinson opened the pull request:
This change is a rework of #22478, originally authored by @vimpunk.

It adds parsing of CSS in parallel with the main script thread. The
big idea here is that when the transfer of stylesheet bytes is
finished, the actual parsing is pushed to a worker thread from the Stylo
thread pool. This also applies for subsequent loads triggered by
`@import` statements.

The design is quite similar to the previous PR with a few significant
changes:

 - Error handling works properly. The `CSSErrorReporter` is a crossbeam
   `Sender` and a `PipelineId` so it can be trivially cloned and sent to
   the worker thread.
 - Generation checking is done both before and after parsing, in order
   to both remove the race condition and avoid extra work when the
   generations do not match.
 - The design is reworked a bit to avoid code duplication, dropping added
   lines from 345 to 160.
 - Now that `process_response_eof` gives up ownership to the
   `FetchResponseListener`, this change avoids all extra copies.

Testing: This shouldn't change observable behavior, so is covered
by existing tests.
Fixes: #20721
Closes: #22478

mrobinson committed:
script: Add support for parsing CSS in parallel

This change is a rework of #22478, originally authored by @vimpunk.

It adds parsing of CSS in parallel with the main script thread. The
big idea here is that when the transfer of stylesheet bytes is
finished, the actual parsing is pushed to a worker thread from the Stylo
thread pool. This also applies for subsequent loads triggered by
`@import` statements.

The design is quite similar to the previous PR with a few significant
changes:

 - Error handling works properly. The `CSSErrorReporter` is a crossbeam
   `Sender` and a `PipelineId` so it can be trivially cloned and sent to
   the worker thread.
 - Generation checking is done both before and after parsing, in order
   to both remove the race condition and avoid extra work when the
   generations do not match.
 - The design is reworked a bit to avoid code duplication, dropping added
   lines from 345 to 160.
 - Now that `process_response_eof` gives up ownership to the
   `FetchResponseListener`, this change avoids all extra copies.

Co-authored-by: mandreyel <mandreyel@protonmail.com>
Signed-off-by: Martin Robinson <mrobinson@igalia.com>

TimvdLippe commented on components/script/stylesheet_loader.rs:
```diff
@@ -119,6 +124,118 @@ impl StylesheetContext {
+            document.invalidate_stylesheets();
+        }
+
+        self.decrement_load_and_render_blockers(owner, &document);
```
If we don't early return above, then why do we decrement twice here?

mrobinson replied:
👍

TimvdLippe commented on components/script/stylesheet_loader.rs:
```diff
@@ -119,6 +124,118 @@ impl StylesheetContext {
+            let event = match any_failed {
+                true => atom!("error"),
+                false => atom!("load"),
+            };
```
Ultra-nit: why not use if instead of match on a boolean?

```suggestion
            let event = if any_failed { atom!("error") } else { atom!("load") };
```

mrobinson replied:
I went with a match because it was one less line, but your approach is even better.

mrobinson replied:
Hrm. It seems that `rustfmt` doesn't like the single-line version, so maybe I'll stick with the `match` since it is one less line than this:

```rust
            let event = match any_failed {
                true => atom!("error"),
                false => atom!("load"),
            };
```

TimvdLippe commented on components/script/stylesheet_loader.rs:
```diff
@@ -150,144 +267,70 @@ impl FetchResponseListener for StylesheetContext {
+                ElementStylesheetLoader::Asynchronous(AsynchronousStylesheetLoader::new(&element))
+            },
+            false => ElementStylesheetLoader::Synchronous { element: &element },
+        };
```
Same ultra-nit about matching booleans:

```suggestion
        let loader = if pref!(dom_parallel_css_parsing_enabled) {
            ElementStylesheetLoader::Asynchronous(AsynchronousStylesheetLoader::new(&element))
        } else {
            ElementStylesheetLoader::Synchronous { element: &element }
        };
```

mrobinson replied:
This makes a lot of sense. I'll make the change.

TimvdLippe reviewed (approved):
Overall this LGTM! I mainly have some nits and clarifying questions about code structure. Glad to see this land 😄

mrobinson commented on components/script/stylesheet_loader.rs:
```diff
@@ -119,6 +124,118 @@ impl StylesheetContext {
+        // From <https://html.spec.whatwg.org/multipage/#link-type-stylesheet>:
+        // > A link element of this type is implicitly potentially render-blocking if the element
+        // > was created by its node document's parser.
+        if matches!(self.source, StylesheetContextSource::LinkElement) && owner.parser_inserted() {
```
Oh, I meant to do this earlier, but I didn't get around to it. I've added an early return here.

mrobinson commented on components/script/stylesheet_loader.rs:
```diff
@@ -119,6 +124,118 @@ impl StylesheetContext {
+            .expect("Stylesheet not loaded by <style> or <link> element!");
+
+        if self.is_for_a_previous_generation(&element) {
+            self.decrement_load_and_render_blockers(owner, &document);
```
Yes! Thank you for spotting this. I meant to do an early return here. This would have caused all kinds of chaos in the future as-is.

mrobinson committed:
Attempt to follow the specification more closely

 - Cancel load blockers when exited early due to generation change
 - Make generation change more general by having the method check
   whether the element still contributes to the styling processing model
   in particular whether it is still connected to the DOM.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson commented:
I have had to make a few minor changes to this PR:

 - Cancel load blockers when exiting early due to generation change
 - Make generation change more general by having the method check
   whether the element still contributes to the styling processing model
   in particular whether it is still connected to the DOM.
